content/advent-2016/how-to-write-good-tests-for-solid-code.md

```diff
@@ -0,0 +1,219 @@
+- Some commands might return some data, but we should be careful, cautious messages that return data and make modifications. We must ensure the changes are never hidden side-effects but required business logic. e.g.: `UserStore.Create(User) (*User, error)` will add a user to the store and must return the information about such user so we can get the ID of the user.
+
+This classification will help us guide what we need to test based on the types of messages affected:
+
+- **Incoming queries**: send the message and assert the response.
+- **Incoming commands**: send the message and assert the public changes. e.g.: call `UserStore.Delete` on an existing user id
+- **Outgoing queries**: nothing to assert.
+- **Outgoing commands**: assert the message sent.
+
+So, how does this apply to our example?
+
+## A bad test
+
+Many people would opt to go straight into developing an integration test for `SignUp`.
+
+```go
+import (
+    "testing"
+
+    "github.com/stretchr/testify/assert"
+)
+
+func TestSignUpSuccessful(t *testing.T) {
+    // 1. Prepare the object with the right context
+    us := newTestUserStore() // user store pointing to a local empty DB
+    un := newTestUserNotifier() // user notifier pointing to a local queue
+    um := NewUserManager(us, un)
+
+    // 2. Send incoming command-query
+    u, err := us.SignUp(User{Email: "ernesto@slackline.io", Password: "secret"})
+
+    // 3. Since it was a query, asssert result
+    assert.NoError(t, err)
+    assert.NotEmpty(t, u.ID)
+    assert.Equal(t, "ernesto@slackline.io", u.email)
+
+    // 4. Since it was a command, assert public side effects
+    stored, err = us.Find(u.ID)
+    assert.NoError(t, err)
+    assert.Equal(t, u.ID, stored.ID)
+    assert.Equal(t, "ernesto@slackline.io", stored.email)
+
+    // 5. testing outgoing messages??
+}
+```
+
+Integration tests provide you end to end checks. But are much more expensive than unit tests:
+
+- Running them is slower since you have to call external services. e.g.: in this case, we have to create/clear our test database for each test.
+- You might not be able to run tests in parallel. e.g.: if you were clearing the same database for each test, running them in parallel could result in false failures due to race conditions preparing the database.
+- With outgoing queries, you need to provision test data before each test. e.g.: if we were to test a `UserManager.Find` method; we would need to add an entry to the `UserStore` before calling `UserManager.Find`.
+- With outgoing commands, you must assert the side effects on your dependency instead of the outgoing commands. e.g.: if you wanted to test `um.notifier.RequestActivation`, you would have to read the queue to make sure `u.ID` got queued up.
+- If your method depends on external services such as third party APIs, they can take seconds to respond and could have downtime. Which would
```

Comment:
Incomplete sentence.

Also, worth mentioning that there might be a cost to the third party APIs? And, imagine sending real emails every time a test were run.